File: surf-api-bukkit/surf-api-bukkit-api/src/main/java/dev/slne/surf/surfapi/bukkit/api/nms/bridges/SurfBukkitNmsNbtBridge.java
package dev.slne.surf.surfapi.bukkit.api.nms.bridges;

import dev.slne.surf.surfapi.bukkit.api.nms.SurfBukkitNmsBridge;
import javax.annotation.ParametersAreNonnullByDefault;
import org.bukkit.entity.EntityType;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.ApiStatus.NonExtendable;
import org.jetbrains.annotations.ApiStatus.ScheduledForRemoval;

@NonExtendable
@ParametersAreNonnullByDefault
public interface SurfBukkitNmsNbtBridge {

  @ScheduledForRemoval(inVersion = "1.21-2.0.0-SNAPSHOT")
  @Deprecated(forRemoval = true)
  default ItemStack makeItemStackEntityInvisible(ItemStack itemStack) {
    return makeItemStackEntityInvisible(itemStack, EntityType.ITEM_FRAME);
  }

  ItemStack makeItemStackEntityInvisible(ItemStack itemStack, EntityType invisibleEntityType);

  /**
   * Get the NBT string from the item stack with the given key
   *
   * @param itemStack the item stack
   * @param key       the key
   * @return the NBT string
   * @deprecated Now only uses the nbt in the custom data components (item stacks now use data
   * components, and nbt should not be avoided)
   */
  @Deprecated
  String getNbtString(ItemStack itemStack, String key);

  static SurfBukkitNmsNbtBridge get() {
    return SurfBukkitNmsBridge.get().getNbtBridge();
  }
}

File: surf-api-bukkit/surf-api-bukkit-api/src/main/kotlin/dev/slne/surf/surfapi/bukkit/api/nms/bridges/SurfBukkitNmsNbtBridge.kt
package dev.slne.surf.surfapi.bukkit.api.nms.bridges

import dev.slne.surf.surfapi.bukkit.api.nms.NmsUseWithCaution
import dev.slne.surf.surfapi.core.api.util.requiredService
import org.bukkit.entity.EntityType
import org.bukkit.inventory.ItemStack

@NmsUseWithCaution
interface SurfBukkitNmsNbtBridge {

    fun makeItemStackEntityInvisible(
        itemStack: ItemStack,
        invisibleEntityType: EntityType
    ): ItemStack

    /**
     * Get the NBT string from the item stack with the given key
     *
     * @param itemStack the item stack
     * @param key       the key
     * @return the NBT string
     */
    @Deprecated("Now only uses the nbt in the custom data components (item stacks now use data components, and nbt should not be avoided)")
    fun getNbtString(itemStack: ItemStack, key: String): String

    companion object {
        val instance = requiredService<SurfBukkitNmsNbtBridge>()
    }
}

@NmsUseWithCaution
val nmsNbtBridge get() = SurfBukkitNmsNbtBridge.instance
